### PR TITLE
Promote Publish CRD OpenAPI to beta

### PR DESCRIPTION
**Pull request by roycaihw**
ref https://github.com/kubernetes/enhancements/issues/692

**Does this PR introduce a user-facing change?**:

```release-note
The CustomResourcePublishOpenAPI feature is now beta and enabled by default. CustomResourceDefinitions with [structural schemas](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190425-structural-openapi.md) now publish schemas in the OpenAPI document served at `/openapi/v2`. CustomResourceDefinitions with non-structural schemas have a `NonStructuralSchema` condition added with details about what needs to be corrected in the validation schema.
```

/sig api-machinery
/cc @sttts 

/hold
pending https://github.com/kubernetes/kubernetes/pull/77554 to be merged

### CODE REVIEW

File: test/e2e/apimachinery/crd_publish_openapi.go [hunk 1]
```diff
@@ -49,7 +49,7 @@ var (
 	metaPattern              = `"kind":"%s","apiVersion":"%s/%s","metadata":{"name":"%s"}`
 )
 
-var _ = SIGDescribe("CustomResourcePublishOpenAPI [Feature:CustomResourcePublishOpenAPI]", func() {
+var _ = SIGDescribe("CustomResourcePublishOpenAPI", func() {
```
Comment (sttts) [inline test/e2e/apimachinery/crd_publish_openapi.go]: Are there any places where the gate is set via the test framework?

Comment (roycaihw) [reply test/e2e/apimachinery/crd_publish_openapi.go]: yes, previously alpha [pull](https://github.com/kubernetes/test-infra/pull/11618/files) and [CI](https://github.com/kubernetes/test-infra/pull/12070/files) jobs looked for regex and enabled `api/all=true`. With this change now the test runs in non-alpha jobs (e.g. [blocking presubmit](https://github.com/kubernetes/test-infra/blob/339ad5a261b51893e26fb62213b9de9d2a9dd97f/config/jobs/kubernetes/sig-gcp/sig-gcp-gce-config.yaml#L21-L58)) which don't modify gates. I'll open a pull to remove the regex pattern from alpha jobs

**Comment by liggitt** [comment]
this looks good to merge from my side. @sttts?

**Comment by liggitt** [comment]
the integration test failures are related. it looks like some test setups do not populate openapiconfig:

https://github.com/kubernetes/kubernetes/blob/b5336132336d9a8696001594b716c85084f98e02/staging/src/k8s.io/apiserver/pkg/server/config.go#L466

so the staticSchema is left nil:

https://github.com/kubernetes/kubernetes/blob/b5336132336d9a8696001594b716c85084f98e02/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L249-L255

so a nil static schema is passed  to the controller:

https://github.com/kubernetes/kubernetes/blob/b5336132336d9a8696001594b716c85084f98e02/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go#L215-L218

and the dereference here panics:

https://github.com/kubernetes/kubernetes/blob/b5336132336d9a8696001594b716c85084f98e02/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/aggregator.go#L23-L26

we have a few options:

1. require openapiconfig, and put a nice warning here to make it clearer to server creators it is required: https://github.com/kubernetes/kubernetes/blob/b5336132336d9a8696001594b716c85084f98e02/staging/src/k8s.io/apiserver/pkg/server/config.go#L432-L438

2. tolerate a nil static schema by replacing it with an empty one here: https://github.com/kubernetes/kubernetes/blob/b5336132336d9a8696001594b716c85084f98e02/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go#L77-L85

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go [hunk 2]
```diff
@@ -203,7 +203,11 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 		crdHandler,
 	)
 	var openapiController *openapicontroller.Controller
-	if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourcePublishOpenAPI) {
+	// OpenAPIVersionedService and StaticOpenAPISpec together serve the /openapi/v2 endpoint on
+	// s.GenericAPIServer. When a generic apiserver is brought up, it may choose to not enable
+	// OpenAPI by having null openAPIConfig, and thus OpenAPIVersionedService and StaticOpenAPISpec
+	// are both null. In that case we don't run the CRD OpenAPI controller.
+	if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourcePublishOpenAPI) && s.GenericAPIServer.OpenAPIVersionedService != nil && s.GenericAPIServer.StaticOpenAPISpec != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go]: OpenAPIVersionedService and StaticOpenAPISpec are populated in PrepareRun(), right? I wouldn't guess they'd be populated here

Comment (roycaihw) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go]: ah that's right. fixing

**Pull request merged**